src/test/java/org/elasticsearch/action/support/replication/ShardReplicationOperationTests.java

```diff
@@ -0,0 +1,443 @@
+        TransportShardReplicationOperationAction<Request, Request, Response>.ReplicationPhase replicationPhase =
+                action.new ReplicationPhase(shardRoutingTable.shardsIt(), request,
+                        new Response(), new ClusterStateObserver(clusterService, logger),
+                        shardRoutingTable.primaryShard(), internalRequest, listener);
+
+        assertThat(replicationPhase.totalShards(), equalTo(shardRoutingTable.size()));
+        assertThat(replicationPhase.pending(), equalTo(assignedReplicas));
+        replicationPhase.start();
+        final CapturingTransport.CapturedRequest[] capturedRequests = transport.capturedRequests();
+        transport.clear();
+        assertThat(capturedRequests.length, equalTo(assignedReplicas));
+        if (assignedReplicas > 0) {
+            assertThat("listener is done, but there are outstanding replicas", listener.isDone(), equalTo(false));
+        }
+        int pending = replicationPhase.pending();
+        int failures = 0;
+        int criticalFailures = 0; // failures that should fail the shard
+        int successfull = 1;
+        for (CapturingTransport.CapturedRequest capturedRequest : capturedRequests) {
+            if (randomBoolean()) {
+                Throwable t;
+                if (randomBoolean()) {
+                    t = new CorruptIndexException("simulated", (String) null);
+                    criticalFailures++;
+                } else {
+                    t = new IndexShardNotStartedException(shardId, IndexShardState.RECOVERING);
+                }
+                failures++;
+                logger.debug("--> simulating failure on {} with [{}]", capturedRequest.node, t.getClass().getSimpleName());
+                transport.handleResponse(capturedRequest.requestId, t);
+            } else {
+                successfull++;
+                transport.handleResponse(capturedRequest.requestId, TransportResponse.Empty.INSTANCE);
+            }
+            pending--;
+            assertThat(replicationPhase.pending(), equalTo(pending));
+            assertThat(replicationPhase.successful(), equalTo(successfull));
+        }
+        assertThat(listener.isDone(), equalTo(true));
+        Response response = listener.get();
+        final ActionWriteResponse.ShardInfo shardInfo = response.getShardInfo();
+        assertThat(shardInfo.getFailed(), equalTo(criticalFailures));
+        assertThat(shardInfo.getFailures(), arrayWithSize(criticalFailures));
+        assertThat(shardInfo.getSuccessful(), equalTo(successfull));
+        assertThat(shardInfo.getTotal(), equalTo(shardRoutingTable.size()));
+
+        assertThat("failed to see enough shard failures", transport.capturedRequests().length, equalTo(criticalFailures));
+        for (CapturingTransport.CapturedRequest capturedRequest : transport.capturedRequests()) {
+            assertThat(capturedRequest.action, equalTo(ShardStateAction.SHARD_FAILED_ACTION_NAME));
+        }
+    }
+
```

Comment:
maybe we add a test that tests what happens if not enough shard copies are available (write consistency)?